# Processing the data (PyTorch)

In [ ]:
# Install the Transformers, Datasets, and Evaluate libraries to run this notebook.
!pip install datasets evaluate transformers[sentencepiece]

Мы будем обучать классификатор последовательностей на одном пакете в PyTorch следующим образом:

In [ ]:
import torch
from transformers import AdamW, AutoTokenizer, AutoModelForSequenceClassification

# Same as before
checkpoint = "bert-base-uncased"
tokenizer = AutoTokenizer.from_pretrained(checkpoint)
model = AutoModelForSequenceClassification.from_pretrained(checkpoint)

In [ ]:
sequences = [
    "I've been waiting for a HuggingFace course my whole life.",
    "This course is amazing!",
]
batch = tokenizer(sequences, padding=True, truncation=True, return_tensors="pt")

# This is new
batch["labels"] = torch.tensor([1, 1])

optimizer = AdamW(model.parameters())
loss = model(**batch).loss
loss.backward()
optimizer.step()

Что делает нижняя часть кода?  
- `batch["labels"] = torch.tensor([1, 1])`	Добавляет метки классов для вычисления loss.  
- `optimizer = AdamW(model.parameters())`	Создаёт оптимизатор для обучения модели.(`model.parameters()` передаёт в оптимизатор все обучаемые параметры модели (веса слоёв BERT).   
- `loss = model(**batch).loss`	Вычисляет функцию потерь (насколько предсказания отличаются от меток).  
- `loss.backward()`	Вычисляет градиенты (как изменить веса, чтобы улучшить предсказания).  
- `optimizer.step()`	Обновляет веса модели (делает её лучше).

📌 Важно: В AutoModelForSequenceClassification функция потерь (loss) автоматически вычисляется, если передать labels в модель.  
🚀 Этот код делает один шаг обучения модели! Если обернуть его в for-цикл по epochs, модель будет тренироваться на всём датасете.

Очевидно, простое обучение модели на двух предложениях не даст хороших результатов. Чтобы получить лучшие результаты, вам нужно будет подготовить больший набор данных.

Будем использовать набор данных MRPC (Microsoft Research Paraphrase Corpus).  Набор данных состоит из 5801 пары предложений с меткой, указывающей, являются ли они парафразами или нет (т. е. означают ли оба предложения одно и то же).

<font color = "lightgreen"> Наборы данных здесь https://huggingface.co/datasets (см. общую документацию здесь https://huggingface.co/docs/datasets/loading). Это один из 10 наборов данных, составляющих бенчмарк GLUE https://gluebenchmark.com/, который является академическим бенчмарком, используемым для измерения производительности моделей машинного обучения в 10 различных задачах классификации текста.</font>  
Загрузим набор данных MRPC:

In [3]:
from datasets import load_dataset

raw_datasets = load_dataset("glue", "mrpc")
raw_datasets

README.md:   0%|          | 0.00/35.3k [00:00<?, ?B/s]

train-00000-of-00001.parquet:   0%|          | 0.00/649k [00:00<?, ?B/s]

validation-00000-of-00001.parquet:   0%|          | 0.00/75.7k [00:00<?, ?B/s]

test-00000-of-00001.parquet:   0%|          | 0.00/308k [00:00<?, ?B/s]

Generating train split:   0%|          | 0/3668 [00:00<?, ? examples/s]

Generating validation split:   0%|          | 0/408 [00:00<?, ? examples/s]

Generating test split:   0%|          | 0/1725 [00:00<?, ? examples/s]

DatasetDict({
    train: Dataset({
        features: ['sentence1', 'sentence2', 'label', 'idx'],
        num_rows: 3668
    })
    validation: Dataset({
        features: ['sentence1', 'sentence2', 'label', 'idx'],
        num_rows: 408
    })
    test: Dataset({
        features: ['sentence1', 'sentence2', 'label', 'idx'],
        num_rows: 1725
    })
})

<font color = "lightgreen">Как вы можете видеть, мы получаем объект DatasetDict, который содержит обучающий набор, проверочный набор и тестовый набор. Каждый из них содержит несколько столбцов (sentence1, sentence2, label и idx) и переменное количество строк, которые являются количеством элементов в каждом наборе (таким образом, в обучающем наборе 3668 пар предложений, в проверочном наборе 408 и в тестовом наборе 1725).

Эта команда загружает и кэширует набор данных по умолчанию в ~/.cache/huggingface/datasets. Вспомните из Главы 2, что вы можете настроить папку кэша, установив переменную среды HF_HOME.

Мы можем получить доступ к каждой паре предложений в нашем объекте raw_datasets с помощью индексации, как в словаре:</font>

In [4]:
raw_train_dataset = raw_datasets["train"]
raw_train_dataset[15]

{'sentence1': 'Rudder was most recently senior vice president for the Developer & Platform Evangelism Business .',
 'sentence2': 'Senior Vice President Eric Rudder , formerly head of the Developer and Platform Evangelism unit , will lead the new entity .',
 'label': 0,
 'idx': 16}

<font color = "lightgreen">Мы видим, что метки уже являются целыми числами, поэтому нам не придется выполнять там предварительную обработку. Чтобы узнать, какое целое число соответствует какой метке, мы можем изучить характеристики нашего raw_train_dataset. Это скажет нам тип каждого столбца:</font>

In [5]:
raw_train_dataset.features

{'sentence1': Value(dtype='string', id=None),
 'sentence2': Value(dtype='string', id=None),
 'label': ClassLabel(names=['not_equivalent', 'equivalent'], id=None),
 'idx': Value(dtype='int32', id=None)}

<font color = "lightgreen">На самом деле label имеет тип ClassLabel, а сопоставление целых чисел имени метки хранится в папке names.  
0 соответствует not_equivalent, а 1 соответствует equal.

<font color = "yellow">Для предварительной обработки набора данных нам нужно преобразовать текст в числа, которые модель может понять, это делается с помощью токенизатора. Мы можем скормить токенизатору одно предложение или список предложений, так что мы можем напрямую токенизировать все первые предложения и все вторые предложения каждой пары следующим образом:

In [6]:
from transformers import AutoTokenizer

checkpoint = "bert-base-uncased"
tokenizer = AutoTokenizer.from_pretrained(checkpoint)
tokenized_sentences_1 = tokenizer(raw_datasets["train"]["sentence1"])
tokenized_sentences_2 = tokenizer(raw_datasets["train"]["sentence2"])

<font color = "lightgreen">Как это выглядит

In [7]:
inputs = tokenizer("This is the first sentence.", "This is the second one.")
inputs

{'input_ids': [101, 2023, 2003, 1996, 2034, 6251, 1012, 102, 2023, 2003, 1996, 2117, 2028, 1012, 102], 'token_type_ids': [0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 1], 'attention_mask': [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1]}

<font color = "yellow">Однако мы не можем просто передать две последовательности в модель и получить прогноз, являются ли эти два предложения парафразами или нет. Нам нужно обработать две последовательности как пару и применить соответствующую предварительную обработку. К счастью, токенизатор также может взять пару последовательностей и подготовить ее так, как ожидает наша модель BERT:

In [8]:
tokenizer.convert_ids_to_tokens(inputs["input_ids"])

['[CLS]',
 'this',
 'is',
 'the',
 'first',
 'sentence',
 '.',
 '[SEP]',
 'this',
 'is',
 'the',
 'second',
 'one',
 '.',
 '[SEP]']

Итак, мы видим, что модель ожидает, что входные данные будут иметь вид [CLS] предложение1 [SEP] предложение2 [SEP], когда есть два предложения. Выравнивание этого с token_type_ids дает нам:  
'''
['[CLS]', 'this', 'is', 'the', 'first', 'sentence', '.', '[SEP]', 'this', 'is', 'the', 'second', 'one', '.', '[SEP]']
[      0,      0,    0,     0,       0,          0,   0,       0,      1,    1,     1,        1,     1,   1,       1]
'''

<small><font color = "lightgreen">Как вы можете видеть, все части ввода, соответствующие [CLS] предложение1 [SEP], имеют идентификатор типа токена 0, в то время как все другие части, соответствующие предложению2 [SEP], имеют идентификатор типа токена 1.
Обратите внимание, что если вы выберете другую контрольную точку, у вас не обязательно будут token_type_ids в ваших токенизированных входных данных (например, они не возвращаются, если вы используете модель DistilBERT). Они возвращаются только тогда, когда модель будет знать, что с ними делать, потому что она видела их во время своего предварительного обучения.
Здесь BERT предварительно обучен с идентификаторами типов токенов, и в дополнение к цели моделирования языка с маской, о которой мы говорили в Главе 1, у него есть дополнительная цель, называемая `прогнозированием следующего предложения`. Цель этой задачи — `смоделировать связь между парами предложений`.
При прогнозировании следующего предложения модели предоставляются пары предложений (со случайно замаскированными токенами) и предлагается предсказать, следует ли второе предложение за первым. Чтобы сделать задачу нетривиальной, половину времени предложения следуют друг за другом в исходном документе, из которого они были извлечены, а другую половину времени два предложения берутся из двух разных документов.
В общем, вам не нужно беспокоиться о том, есть ли token_type_id в ваших токенизированных входных данных: пока вы используете одну и ту же контрольную точку для токенизатора и модели, все будет хорошо, поскольку токенизатор знает, что предоставить своей модели.
Теперь, когда мы увидели, как наш токенизатор может работать с одной парой предложений, мы можем использовать его для токенизации всего нашего набора данных: как и в предыдущей главе, мы можем передать токенизатору список пар предложений, передав ему список первых предложений, а затем список вторых предложений. Это также совместимо с параметрами заполнения и усечения, которые мы видели в Главе 2. Итак, один из способов предварительной обработки обучающего набора данных:</small>

In [9]:
tokenized_dataset = tokenizer(
    raw_datasets["train"]["sentence1"],
    raw_datasets["train"]["sentence2"],
    padding=True,
    truncation=True,
)

<small>Это работает хорошо, но у него есть недостаток в том, что возвращается словарь (с нашими ключами, input_ids, awareness_mask и token_type_ids, а также значениями, которые являются списками списков). Это также будет работать только в том случае, если у вас достаточно оперативной памяти для хранения всего набора данных во время токенизации (тогда как наборы данных из библиотеки 🤗 Datasets представляют собой файлы Apache Arrow, хранящиеся на диске, поэтому вы сохраняете только запрошенные вами образцы, загруженные в память).

Чтобы сохранить данные как набор данных, мы будем использовать метод Dataset.map(). Это также дает нам некоторую дополнительную гибкость, если нам нужно больше предварительной обработки, чем просто токенизация. Метод map() работает, применяя функцию к каждому элементу набора данных, поэтому давайте определим функцию, которая токенизирует наши входные данные:</small>

In [10]:
def tokenize_function(example):
    return tokenizer(example["sentence1"], example["sentence2"], truncation=True)

<small><font color='yellow'>Эта функция принимает словарь (например, элементы нашего набора данных) и возвращает новый словарь с ключами input_ids, awareness_mask и token_type_ids. Обратите внимание, что это также работает, если словарь-пример содержит несколько образцов (каждый ключ как список предложений), поскольку токенизатор работает со списками пар предложений, как было показано ранее. Это позволит нам использовать опцию batched=True в нашем вызове map(), что значительно ускорит токенизацию. Токенизатор поддерживается токенизатором, написанным на Rust из библиотеки 🤗 Tokenizers. Этот токенизатор может быть очень быстрым, но только если мы дадим ему много входных данных одновременно.
Обратите внимание, что мы пока оставили аргумент padding в нашей функции токенизации. Это связано с тем, что заполнение всех образцов до максимальной длины неэффективно: лучше заполнять образцы, когда мы создаем пакет, так как тогда нам нужно заполнить только до максимальной длины в этом пакете, а не до максимальной длины во всем наборе данных. Это может сэкономить много времени и вычислительной мощности, когда входные данные имеют очень изменчивую длину!
Вот как мы применяем функцию токенизации ко всем нашим наборам данных одновременно. Мы используем batched=True в нашем вызове map, поэтому функция применяется к нескольким элементам нашего набора данных одновременно, а не к каждому элементу отдельно. Это позволяет ускорить предварительную обработку.</small>

In [11]:
tokenized_datasets = raw_datasets.map(tokenize_function, batched=True)
tokenized_datasets

Map:   0%|          | 0/3668 [00:00<?, ? examples/s]

Map:   0%|          | 0/408 [00:00<?, ? examples/s]

Map:   0%|          | 0/1725 [00:00<?, ? examples/s]

DatasetDict({
    train: Dataset({
        features: ['sentence1', 'sentence2', 'label', 'idx', 'input_ids', 'token_type_ids', 'attention_mask'],
        num_rows: 3668
    })
    validation: Dataset({
        features: ['sentence1', 'sentence2', 'label', 'idx', 'input_ids', 'token_type_ids', 'attention_mask'],
        num_rows: 408
    })
    test: Dataset({
        features: ['sentence1', 'sentence2', 'label', 'idx', 'input_ids', 'token_type_ids', 'attention_mask'],
        num_rows: 1725
    })
})

Библиотека 🤗 Datasets применяет эту обработку, добавляя новые поля в наборы данных, по одному для каждого ключа в словаре, возвращаемом функцией предварительной обработки:

In [ ]:
tokenized_datasets['train'][15]

<small>Вы даже можете использовать многопроцессорную обработку при применении функции предварительной обработки с map(), передав аргумент num_proc. Мы не сделали этого здесь, потому что библиотека 🤗 Tokenizers уже использует несколько потоков для более быстрой токенизации наших образцов, но если вы не используете быстрый токенизатор, поддерживаемый этой библиотекой, это может ускорить вашу предварительную обработку.

Наша tokenize_function возвращает словарь с ключами input_ids, awareness_mask и token_type_ids, поэтому эти три поля добавляются во все разделы нашего набора данных. Обратите внимание, что мы также могли бы изменить существующие поля, если бы наша функция предварительной обработки возвращала новое значение для существующего ключа в наборе данных, к которому мы применили map().

Последнее, что нам нужно будет сделать, это дополнить все примеры до длины самого длинного элемента, когда мы объединяем элементы вместе — метод, который мы называем динамическим заполнением.</small>

## Dynamic padding (динамическое заполнение)

Функция, которая отвечает за объединение образцов внутри пакета, называется функцией collate. Это аргумент, который вы можете передать при создании DataLoader, по умолчанию это функция, которая просто преобразует ваши образцы в тензоры PyTorch и объединяет их (рекурсивно, если ваши элементы являются списками, кортежами или словарями). В нашем случае это будет невозможно, поскольку входные данные, которые у нас есть, не будут иметь одинаковый размер. Мы намеренно отложили заполнение, чтобы применять его только по мере необходимости к каждому пакету и избегать слишком длинных входных данных с большим количеством заполнения. Это значительно ускорит обучение, но учтите, что если вы обучаетесь на TPU, это может вызвать проблемы — TPU предпочитают фиксированные формы, даже если для этого требуется дополнительное заполнение.

Чтобы сделать это на практике, нам нужно определить функцию collate, которая будет применять правильное количество заполнения к элементам набора данных, которые мы хотим объединить в пакет. К счастью, библиотека 🤗 Transformers предоставляет нам такую ​​функцию через DataCollatorWithPadding. Она принимает токенизатор, когда вы создаете ее экземпляр (чтобы знать, какой токен заполнения использовать и ожидает ли модель, что заполнение будет слева или справа от входов) и сделает все, что вам нужно:

In [13]:
from transformers import DataCollatorWithPadding

data_collator = DataCollatorWithPadding(tokenizer=tokenizer)

Чтобы протестировать эту новую игрушку, давайте возьмем несколько образцов из нашего обучающего набора, которые мы хотели бы объединить в пакет. Здесь мы удалим столбцы idx, sentence1 и sentence2, поскольку они не понадобятся и содержат строки (и мы не можем создавать тензоры со строками), и посмотрим на длину каждой записи в пакете:

In [15]:
samples = tokenized_datasets["train"][:8]
samples = {k: v for k, v in samples.items() if k not in ["idx", "sentence1", "sentence2"]}
[len(x) for x in samples["input_ids"]]

[50, 59, 47, 67, 59, 50, 62, 32]

Неудивительно, что мы получаем образцы разной длины, от 32 до 67. Динамическое заполнение означает, что образцы в этом пакете должны быть дополнены до длины 67, максимальной длины внутри пакета. Без динамического заполнения все образцы должны быть дополнены до максимальной длины во всем наборе данных или максимальной длины, которую может принять модель. Дополним:

In [16]:
batch = data_collator(samples)
{k: v.shape for k, v in batch.items()}

{'input_ids': torch.Size([8, 67]),
 'token_type_ids': torch.Size([8, 67]),
 'attention_mask': torch.Size([8, 67]),
 'labels': torch.Size([8])}

Выглядит хорошо! Теперь, когда мы перешли от сырого текста к пакетам, с которыми может работать наша модель, мы готовы к ее тонкой настройке!

✏️ Попробуйте! Повторите предварительную обработку на наборе данных GLUE SST-2. Это немного отличается, так как состоит из отдельных предложений, а не пар, но все остальное, что мы сделали, должно выглядеть так же. Для более сложной задачи попробуйте написать функцию предварительной обработки, которая работает с любой из задач GLUE.

____________________________________________________

In [23]:
raw_datasets = load_dataset("gimmaru/glue-sst2")
raw_datasets

README.md:   0%|          | 0.00/775 [00:00<?, ?B/s]

(…)-00000-of-00001-02c8af09f7a87704.parquet:   0%|          | 0.00/72.7k [00:00<?, ?B/s]

Generating validation split:   0%|          | 0/872 [00:00<?, ? examples/s]

DatasetDict({
    validation: Dataset({
        features: ['sentence', 'label', 'idx'],
        num_rows: 872
    })
})

In [24]:
raw_train_dataset = raw_datasets["validation"]
raw_train_dataset[40]

{'sentence': "though it 's become almost redundant to say so , major kudos go to leigh for actually casting people who look working-class . ",
 'label': 1,
 'idx': 83}

In [25]:
raw_datasets_2 = load_dataset("Tohrumi/glue_sst2_10k")
raw_datasets_2

README.md:   0%|          | 0.00/654 [00:00<?, ?B/s]

train-00000-of-00001.parquet:   0%|          | 0.00/459k [00:00<?, ?B/s]

eval-00000-of-00001.parquet:   0%|          | 0.00/27.1k [00:00<?, ?B/s]

test-00000-of-00001.parquet:   0%|          | 0.00/42.4k [00:00<?, ?B/s]

Generating train split:   0%|          | 0/10000 [00:00<?, ? examples/s]

Generating eval split:   0%|          | 0/500 [00:00<?, ? examples/s]

Generating test split:   0%|          | 0/500 [00:00<?, ? examples/s]

DatasetDict({
    train: Dataset({
        features: ['sentence', 'label', 'idx'],
        num_rows: 10000
    })
    eval: Dataset({
        features: ['sentence', 'label', 'idx'],
        num_rows: 500
    })
    test: Dataset({
        features: ['sentence', 'label', 'idx'],
        num_rows: 500
    })
})

In [30]:
raw_train_dataset_2 = raw_datasets_2["train"]
raw_train_dataset_2[704]

{'sentence': 'has been , pardon the pun , sucked out and replaced by goth goofiness . ',
 'label': 0,
 'idx': 5017}